Query debugging tracer

jtao15 opened the pull request:
Add query debugging tracer for Presto as https://github.com/prestosql/presto/issues/1826
This pr includes tracer creation and propagations and some basic tracer events.

## Tracer high level design
### Objectives
- Develop an extensible tracer framework that allows Presto engine to emit tracing events during query execution
- Implement propagation of tracer to connectors to enable publishing events from within connectors
- Add useful debugging events to the aforementioned framework

### Key design points
- Tracers track interesting execution actions per query
- Context dependency
Tracers keep the execution contexts tracked, such as `queryId`, `stageId`, `taskId`, `worker nodeId`, etc. such that tracers events can be uniquely identified.
- Introduction of tracers
`EventListenerManger` is injected in tracer as an event consumer (emitter) when the tracer is created by `TracerFactroy`.
Tracer defined in SPI package are wrapped in `ConnectorEmitter` such that `ConnectorTracer` can be created with emitters without explicitly referencing an engine tracer.

     ![Untitled Diagram (3)](https://user-images.githubusercontent.com/50065407/70290677-822c7280-178d-11ea-97c4-0664a8f2f273.jpg)

- Hierarchical
As query executes hierarchically, tracers are created and propagated correspondingly. QueryTracer is created before the query is planned on coordinator. `StageTracers` are created based on the context of their parent `QueryTracer` with stageId assigned. `TaskTracers` are then created following the parent StageTracers. 
On workers, `TaskTracers` are created from scratch with task context (queryId, stageId, taskId) assigned. `DriverTracers` and `OperatorTracers` are created accordingly as the execution propagates.

     ![TracerPropagation (6)](https://user-images.githubusercontent.com/50065407/70290763-cb7cc200-178d-11ea-9cd9-6af3ef740cd8.jpg)

- `ConnectorTracers` are created by `ConnectorTracerFactory` with `ConnectorTracerEmitter`(wrapper of engine tracer). In connectors the `ConnectorTracer` is propagated through connector context, the context is extensible to include other information for future development so there will not be an interface change.

- Tracer events are semi-structured. Structured fields include execution object ids, event type, worker node id, timestamp, etc. Other information is kept as generic payload(implementation yet to be decided).
- `TracerEventType` enums and `ConnectorEventType` enums share the same interface: `EventTypeSupplier`, so that connectors are able to define their own event types with connector name prefix.


### Proposed events details

| Coordinator/Worker | EngineTracer/ConnectorTracer | Execution unit | Event                                                      | payload (default as null)                                                                 | Count for sample query(SELECT * FROM hive.u_name.test_table)  | Possible multiple occurence per execution unit | test                          |
|--------------------|------------------------------|----------------|------------------------------------------------------------|-------------------------------------------------------------------------------------------|---------------------------------------------------------------|------------------------------------------------|-------------------------------|
| Coordinator        | EngineTracer                 | Query          | PLAN_QUERY_START                                           |                                                                                           | 1                                                             |                                                | TestTracer                    |
|                    |                              |                | PLAN_QUERY_END                                             |                                                                                           | 1                                                             |                                                | TestTracer                    |
|                    |                              |                | QUERY_STATE_CHANGE_EVENTS(QUEUED, PLANNING, STARTING etc.  |                                                                                           | 1                                                             |                                                |                               |
|                    |                              | Stage          | STAGE_STATE_CHANGE_EVENT(PLANNED, SCHEDULING, RUNNING etc. |                                                                                           | 2                                                             |                                                | TestSqlStageExecution         |
|                    |                              | Task           | SCHEDULE_TASK_WITH_SPLITS                                  | { "splits": splits( Multimap<PlanNodeId, Split>) }                                        | 2                                                             |                                                | TestSqlStageExecution         |
|                    |                              |                | ADD_SPLITS_TO_TASK                                         | { "splits": splits( Multimap<PlanNodeId, Split>) }                                        | 0                                                             | TRUE                                           |                               |
|                    |                              |                | SEND_UPDATE_TASK_REQUEST_START                             | { "request": request(Request.toString( )) }                                               | 5                                                             | TRUE                                           | TestHttpRemoteTask            |
|                    |                              |                | SEND_UPDATE_TASK_REQUEST_END                               | { "response": value(TaskInfo.toString( ))} or { "error": cause(Throwable.toString( ) )}   | 5                                                             | TRUE                                           | TestHttpRemoteTask            |
|                    | ConnectorTracer              |                | Hive.LIST_FILE_STATES_START                                | { "path": path(Path.toString( ))}                                                         | 1                                                             |                                                | TestBackgroundHiveSplitLoader |
|                    |                              |                | Hive.LIST_FILE_STATES_END                                  | { "path": path(Path.toString( ))}                                                         | 1                                                             |                                                | TestBackgroundHiveSplitLoader |
|                    |                              |                | Hive.ORC_OPEN_FILE_START                                   |                                                                                           | 0                                                             |                                                |                               |
|                    |                              |                | Hive.ORC_OPEN_FILE_END                                     |                                                                                           | 0                                                             |                                                |                               |
| Worker             | EngineTracer                 | Task           | CREATE_LOCAL_PLAN_START                                    |                                                                                           | 2                                                             |                                                |                               |
|                    |                              |                | CREATE_LOCAL_PLAN_END                                      |                                                                                           | 2                                                             |                                                |                               |
|                    |                              |                | TASK_STATE_CHANGE_EVENTS(PLANNED, RUNNING, FINISHED etc.   |                                                                                           | 2                                                             |                                                | TestSqlTask                   |
|                    |                              | Split/Driver   | CREATED_DRIVER                                             |                                                                                           | 18                                                            |                                                |                               |
|                    |                              |                | SPLIT_DRIVER_CREATED                                       |                                                                                           | 18                                                            |                                                | TestTaskExecutor              |
|                    |                              |                | SPLIT_ADDED                                                |                                                                                           |                                                               |                                                |                               |
|                    |                              |                | SPLIT_SCHEDULED                                            |                                                                                           | 40                                                            | TRUE                                           | TestTaskExecutor              |
|                    |                              |                | SPLIT_UNSCHEDULED                                          |                                                                                           |                                                               |                                                |                               |
|                    |                              |                | SPLIT_STARTS_WAITING                                       |                                                                                           | 2                                                             | TRUE                                           | TestTaskExecutor              |
|                    |                              |                | SPLIT_ENDS_WAITING                                         |                                                                                           |                                                               |                                                |                               |
|                    |                              |                | SPLIT_BLOCKED                                              |                                                                                           | 18                                                            | TRUE                                           |                               |
|                    |                              |                | SPLIT_UNBLOCKED                                            |                                                                                           | 20                                                            | TRUE                                           |                               |
|                    |                              |                | SPLIT_DESTROY_INVOKED                                      |                                                                                           | 18                                                            |                                                | TestTaskExecutor              |
|                    |                              |                | SPLIT_FINISHED                                             |                                                                                           | 17                                                            |                                                | TestTaskExecutor              |
|                    |                              | Operator       | ADD_SPLIT_TO_OPERATOR                                      |                                                                                           | 2                                                             |                                                | TestDriver                    |
|                    | ConnectorTracer              |                | Orc.READ_ORC_TAIL_START                                    | { "path": path(orcDataSource.getId().toString()) }                                        | 2                                                             |                                                | TestOrcReaderPositions        |
|                    |                              |                | Orc.READ_ORC_TAIL_END                                      | { "path": path(orcDataSource.getId().toString()), "orc tail": buffer(Slice.toString( )) } | 2                                                             |                                                | TestOrcReaderPositions        |
|                    |                              |                | Orc.READ_STRIPE_START                                      | { "stripe": stripeInformation(StripeInformation.toString( )) }                            | 2                                                             |                                                | TestOrcReaderPositions        |
|                    |                              |                | Orc.READ_STRIPE_END                                        | { "stripe": stripeInformation(StripeInformation.toString( )) }                            | 2                                                             |                                                | TestOrcReaderPositions        |
|                    |                              |                | Orc.READ_ORC_COMPLETE_FOOTER_START                         |                                                                                           | 0                                                             |                                                |                               |
|                    |                              |                | Orc.READ_ORC_COMPLETE_FOOTER_END                           | { "orc footer": completeFoorterSlice(Slice.toString( )) }                                 | 0                                                             |                                                |                               |
|                    |                              |                | Orc.READ_BLOCK_START                                       |                                                                                           | 2                                                             |                                                | TestOrcReaderPositions        |
|                    |                              |                | Orc.READ_BLOCK_END                                         | { "block": block(Block.toString()) }                                                      | 2                                                             |                                                | TestOrcReaderPositions        |

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/Tracer.java:
```diff
@@ -0,0 +1,259 @@
+    private final Optional<String> operatorId;
+    private final Optional<String> nodeId;
+
+    private Tracer(Builder builder)
```
It's better to have a `build` method inside the Builder class that constructs Tracer object with member variables as parameters.

jtao15 commented on presto-main/src/test/java/io/prestosql/operator/TestDriver.java:
```diff
@@ -502,9 +538,23 @@ public NotBlockedTableScanOperator(
+        private final String mockProperty;
+
+        @JsonCreator
+        public MockSplit(@JsonProperty("mockProperty") String mockProperty)
```
JsonCodec throws errors when it converts an empty split to json payload, so I added a dummy field here.

jtao15 commented on presto-main/src/main/java/io/prestosql/operator/TableScanOperator.java:
```diff
@@ -280,13 +299,32 @@ public Page getOutput()
             page = page.getLoadedPage();
+            if (tracer.isEnabled()) {
+                Map<String, Object> payload = new HashMap<>();
+                payload.put("page", page.toString());
```
I think the page.toString() will not include the extra data, and we may be interested which page it is when an event listener receive a `PageLoaded` event.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java:
```diff
@@ -101,7 +110,17 @@ protected LocatedFileStatus computeNext()
+                Map<String, Object> map = new HashMap<>();
+                map.put("path", path.toString());
+                payload = jsonCodec.toJson(map);
+                tracer.emitEvent(ConnectorTracerActionType.LIST_FILE_STATES_START, payload);
```
I think this can be compacted: 
```
if (tracer.isEnabled()) {
    tracer.event(LIST_FILE_STATS_START, toJson(ImmutableMap.of("path", path));
}
```

also applicable to other emissions

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -399,7 +448,18 @@ private void invokeNoMoreSplitsIfNecessary()
+                tracer.emitEvent(ConnectorTracerActionType.LOAD_SPLIT_START, payload);
+            }
             Optional<InternalHiveSplit> internalHiveSplit = splitFactory.createInternalHiveSplit((FileSplit) inputSplit);
+            tracer.emitEvent(ConnectorTracerActionType.LOAD_SPLIT_END, payload);
```
`isEnabled` for uniformity, it won't matter in terms of correctness. Thinking if we could do something better instead of double checking for  `isEnabled`.

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,123 @@
+        this.operatorId = operatorId;
+        this.nodeId = nodeId;
+        this.threadName = requireNonNull(threadName, "threadName is null");
+        this.timestampEpochMillis = timeMillis;
```
@jtao15

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,151 @@
+        this.splitId = splitId;
+        this.operatorId = operatorId;
+        this.nodeId = nodeId;
+        this.host = host;
```
@jtao15 nullity?

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,151 @@
+        return "queryId: " + queryId +
+                ", stageId: " + (stageId == null ? "" : stageId) +
+                ", taskId: " + (taskId == null ? "" : taskId) +
+                ", PipelineId: " + (pipelineId == null ? "" : pipelineId) +
```
let's be consistent about the capitalisation for all fields

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,240 @@
+    private final Optional<String> splitId;
+    private final Optional<String> operatorId;
+    // presto worker node emitting tracer events
+    private final Optional<String> nodeId;
```
@jtao15

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,297 @@
+    {
+        private final boolean enabled;
+        private final Consumer<TracerEvent> emitter;
+        private String queryId;
```
final

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,297 @@
+
+        public Builder withStageId(String stageId)
+        {
+            requireNonNull(stageId);
```
add message, same in other places

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,297 @@
+            Optional<String> splitId,
+            Optional<String> operatorId,
+            Optional<String> nodeId,
+            Optional<String> host,
```
let's combine host and port in the same string, this information is not heirarchical

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,297 @@
+
+        public Builder(Consumer<TracerEvent> emitter, String queryId, boolean isEnabled)
+        {
+            this.emitter = requireNonNull(emitter);
```
message

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/NoOpTracer.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public void emitEvent(TracerEventTypeSupplier tracerEventTypeSupplier, PayloadBuilder payloadBuilder)
+    { }
```
put in previous line

jtao15 replied:
Seems like the style check won't let me to move this to previous line

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/tracer/PayloadBuilder.java:
```diff
@@ -0,0 +1,19 @@
+ */
+package io.prestosql.spi.tracer;
+
+public interface PayloadBuilder
```
add comment on why this interface is required

phd3 commented on presto-tests/src/test/java/io/prestosql/execution/TestEventListener.java:
```diff
@@ -278,6 +289,7 @@ public synchronized void initialize(int numEvents)
             queryCreatedEvents = ImmutableList.builder();
             queryCompletedEvents = ImmutableList.builder();
             splitCompletedEvents = ImmutableList.builder();
+            tracerEvents = Collections.synchronizedList(new ArrayList<>());
```
why is this different, and not an immutable list builder?

phd3 commented on presto-main/src/test/java/io/prestosql/tracer/TestTracer.java:
```diff
@@ -0,0 +1,75 @@
+    private static final String TASK_ID = "task";
+    private static final String NODE_ID = "node";
+    private static final TracerEventType QUERY_ACTION = PLAN_QUERY_START;
+    private static final TracerEventType TASK_ACTION = PLAN_QUERY_END;
```
rename?

phd3 commented on presto-main/src/test/java/io/prestosql/tracer/TestTracer.java:
```diff
@@ -0,0 +1,75 @@
+    @Test
+    public void testTracer()
+    {
+        ImmutableList.Builder<TracerEvent> tracerEvents = new ImmutableList.Builder<>();
```
= ImmutableList.builder()

phd3 commented on presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQueryFactory.java:
```diff
@@ -92,14 +94,16 @@ public DispatchQuery createDispatchQuery(
+            Tracer tracer)
     {
         WarningCollector warningCollector = warningCollectorFactory.create();
+        URI queryLocation = locationFactory.createQueryLocation(session.getQueryId());
```
discussed offline, looks like using NodeInfo might be a better choice to get host info

phd3 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -418,6 +424,7 @@ public void testComplex(PipelineExecutionStrategy executionStrategy)
                     ImmutableList.of(scan2NodeId, scan0NodeId),
                     executionStrategy == GROUPED_EXECUTION ? StageExecutionDescriptor.fixedLifespanScheduleGroupedExecution(ImmutableList.of(scan0NodeId, scan2NodeId)) : StageExecutionDescriptor.ungroupedExecution());
             TaskContext taskContext = newTestingTaskContext(taskNotificationExecutor, driverYieldExecutor, taskStateMachine);
+
```
unrelated change, please revert

phd3 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -131,7 +131,10 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
+            final String queryId = "0";
+            final String stageId = "1";
+            final String taskId = "2";
+            TaskStateMachine taskStateMachine = new TaskStateMachine(TaskId.valueOf(queryId + "." + stageId + "." + taskId), taskNotificationExecutor);
```
I don't think this is used

jtao15 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlStageExecution.java:
```diff
@@ -99,6 +105,8 @@ private void testFinalStageInfoInternal()
         NodeTaskMap nodeTaskMap = new NodeTaskMap(new FinalizerService());
 
         StageId stageId = new StageId(new QueryId("query"), 0);
+        List<TracerEvent> tracerEvents = Collections.synchronizedList(new ArrayList<>());
```
Noop tracer will not emit events, while in the test we want to collect emitted events.

jtao15 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTask.java:
```diff
@@ -324,6 +353,48 @@ private SqlTask createInitialTask()
+    {
+        try {
+            // wait a second for events to propagate
+            Thread.sleep(1000);
```
The event emission shouldn't be blocked for testing tracer, we need to wait until all events are collected.

jtao15 commented on presto-orc/src/main/java/io/prestosql/orc/OrcBlockFactory.java:
```diff
@@ -76,10 +82,15 @@ public final Block load()
                 if (loadFully) {
                     block = block.getLoadedBlock();
                 }
+                if (block != null) {
```
I've removed the check. `BlockInfo` is needed here since only final fields can be referenced in lamda expression.

jtao15 commented on presto-orc/src/main/java/io/prestosql/orc/OrcTracerEventType.java:
```diff
@@ -0,0 +1,35 @@
+
+import io.prestosql.spi.tracer.TracerEventTypeSupplier;
+
+public enum OrcTracerEventType
```
I think presto-hive is not a dependency for presto-orc, we cannot access to these event types if put in presto-hive.

jtao15 commented on presto-orc/pom.xml:
```diff
@@ -61,6 +61,11 @@
             <artifactId>stats</artifactId>
         </dependency>
 
+        <dependency>
```
The pros to keep jsonCodec in TracerFactory is we don’t need to create a jsonCodec object whenever an event is emitted. If we use jsonCodec for operator level events it may take lots of resources for large queries. Also we will not introduce the dependencies to connector packages which don’t use jsonCodec currently.
The cons would be it makes it a little bit hard to read. 
I've updated this by creating jsonCodec object in TracerFactory for now.

jtao15 commented:
Thanks for the comments @phd3 , I've addressed them accordingly.
Appreciated the suggestions we discussed offline @martint , I incorporated them in this pr also. Feel free to drop any comments or feedbacks. Thank you!

martint commented on presto-main/src/main/java/io/prestosql/tracer/DefaultTracerFactory.java:
```diff
@@ -0,0 +1,54 @@
+    private final EventListenerManager eventListenerManager;
+    private final NodeInfo nodeInfo;
+    private final URI uri;
+    private final JsonCodec<Map<String, Object>> jsonCodec = JsonCodec.mapJsonCodec(String.class, Object.class);
```
Inject this codec and register it in the Guice module with `jsonCodecBinder`. Otherwise, it gets initialized with whatever defaults Jackson picks vs the curated options and modules configured by airlift's `ObjectMapperProvider`.

martint commented on presto-main/src/main/java/io/prestosql/tracer/DefaultTracerFactory.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public Tracer createBasicTracer(String queryId, Session session)
+    {
+        return DefaultTracer.createBasicTracer(event -> eventListenerManager.tracerEventOccurred(event), payloadMap -> jsonCodec.toJson(payloadMap), nodeInfo.getNodeId(), uri, queryId, isQueryDebuggingTracerEnabled(session));
```
This line is too long. Place each argument on a separate line to make it easier to read. Also, you can use method references for the inline lambdas:

```java
return DefaultTracer.createBasicTracer(
        eventListenerManager::tracerEventOccurred,
        jsonCodec::toJson,
        nodeInfo.getNodeId(),
        uri,
        queryId,
        isQueryDebuggingTracerEnabled(session));
```

martint replied:
Instead of passing the `enabled` flag to `DefaultTracer.createBasicTracer`, why not return a `NoOpTracer` instead?

martint commented on presto-main/src/main/java/io/prestosql/tracer/NoOpTracerFactory.java:
```diff
@@ -0,0 +1,36 @@
+public class NoOpTracerFactory
+        implements TracerFactory
+{
+    public NoOpTracerFactory()
```
Default constructor is not necessary.

martint commented on presto-main/src/main/java/io/prestosql/tracer/TracerFactory.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface TracerFactory
+{
+    Tracer createBasicTracer(String queryId, Session session);
```
What's a "basic" tracer? Maybe just rename to `createTracer` or `newTracer`?

martint commented on presto-main/src/test/java/io/prestosql/tracer/TestTracer.java:
```diff
@@ -0,0 +1,85 @@
+        Tracer taskTracer = queryTracer.newTracerWithStageId(STAGE_ID).newTracerWithTaskId(TASK_ID);
+        taskTracer.emitEvent(TASK_EVENT, () -> PAYLOAD);
+
+        ImmutableList<TracerEvent> events = tracerEvents.build();
```
Use `List<TracerEvent` in type declaration.

martint commented on presto-main/src/test/java/io/prestosql/tracer/TestTracer.java:
```diff
@@ -0,0 +1,85 @@
+    public void testTracer()
+    {
+        ImmutableList.Builder<TracerEvent> tracerEvents = new ImmutableList.Builder();
+        Tracer queryTracer = DefaultTracer.createBasicTracer(event -> tracerEvents.add(event), map -> jsonCodec.toJson(map), NODE_ID, TEST_URI, QUERY_ID, true);
```
Use method references for inline lambdas.

martint commented on presto-main/src/test/java/io/prestosql/tracer/TestTracer.java:
```diff
@@ -0,0 +1,85 @@
+    private static final TracerEventType TASK_EVENT = PLAN_QUERY_END;
+    private static final Map<String, Object> PAYLOAD = ImmutableMap.of("key", "value");
+
+    private final JsonCodec<Map<String, Object>> jsonCodec = JsonCodec.mapJsonCodec(String.class, Object.class);
```
Consider making these variables in the `testTracer` method. There's no real reason for them to be class-level constants.

martint commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,146 @@
+
+    private final String threadName;
+
+    //time of the source node
```
What's the "source node"?

jtao15 replied:
I mean the node where the tracer events are sent, either the coordinator or a worker.

martint commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,146 @@
+    private final String threadName;
+
+    //time of the source node
+    private final long timestampEpochMillis;
```
Use `Instant` instead.

martint commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,265 @@
+import static java.util.Objects.requireNonNull;
+
+// Tracer keeps track of Presto execution unit contexts and sends debugging purpose events via its emitter.
+// Users should not build complex metrics based on the action events.
```
What do you mean by "Users should not build complex metrics based on the action events." ?

jtao15 replied:
The tracer events should be used for developing debugging or diagnostic purposes only, and not be used to build metrics for production.

martint commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,265 @@
+
+    public static Tracer createBasicTracer(Consumer<TracerEvent> emitter, Function<Map<String, Object>, String> payloadEncoder, String nodeId, URI uri, String queryId, boolean enabled)
+    {
+        return (new Builder(emitter, payloadEncoder, nodeId, uri, queryId, enabled)).build();
```
Parenthesis around `new Builder` are unnecessary. Consider using a factory method in the builder to improve readability:

```
builder(...).build()
```

martint commented on presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java:
```diff
@@ -0,0 +1,265 @@
+    @Override
+    public Tracer newTracerWithStageId(String stageId)
+    {
+        return (new Builder(this)).withStageId(stageId).build();
```
Same as above. Consider rewriting as follows by adding a `builder()` factory method:

```java
builder(this).withStageId(stageId).build();
```

martint commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,146 @@
+public class TracerEvent
+{
+    private final String nodeId;
+    private final String uri;
```
What URI does this represent?

jtao15 replied:
It's the URI of the node where the tracer events are sent.

martint commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/TracerEvent.java:
```diff
@@ -0,0 +1,146 @@
+    private final String taskId;
+    private final String pipelineId;
+    private final String splitId;
+    private final String operatorId;
```
Use `Optional` for any field that may be omitted. It makes it easier for users of the API to understand its semantics.

martint commented on presto-spi/src/main/java/io/prestosql/spi/tracer/Tracer.java:
```diff
@@ -0,0 +1,31 @@
+
+public interface Tracer
+{
+    Tracer newTracerWithStageId(String stageId);
```
I'd remove the `newTracer` part from the names of these methods and leave them as `withStageId`, `withTaskId`, etc.

martint commented on presto-spi/src/main/java/io/prestosql/spi/tracer/Tracer.java:
```diff
@@ -0,0 +1,31 @@
+
+    void emitEvent(TracerEventTypeSupplier eventTypeSupplier, PayloadBuilder payloadBuilder);
+
+    boolean isEnabled();
```
See my comment elsewhere about instantiating a NoOpTracer. In that case, this flag is unnecessary -- a tracer is always enabled, but may do nothing.

martint commented on presto-spi/src/main/java/io/prestosql/spi/eventlistener/EventListener.java:
```diff
@@ -26,4 +26,8 @@ default void queryCompleted(QueryCompletedEvent queryCompletedEvent)
     {
     }
+
+    default void tracerEventOccurred(TracerEvent tracerEvent)
```
This method name sounds funny (I don't have any suggestions for alternate names, yet -- let me think about it).

phd3 commented:
@jtao15 can you please fix the failing test?

tooptoop4 commented:
@jtao15 conflicts

RosterIn commented:
@jtao15 This can be awesome!!! is there any chance you will continue this?

bitsondatadev commented:
:wave: @jtao15 - this PR is inactive and doesn't seem to be under development, and it might already be implemented. If you'd like to continue work on this at any point in the future, feel free to re-open.
